runtime/oti/ContinuationHelpers.hpp

```diff
@@ -323,6 +323,74 @@ class VM_ContinuationHelpers {
 
 		return foundInList;
 	}
+
+	/**
+	 * Logic to notify virtual threads waiting on an object monitor.
+	 *
+	 * @param[in] vmThread the J9VMThread
+	 * @param[in] objectMonitor the object monitor on which the virtual threads are waiting
+	 * @param[in] notifyAll indicates if all virtual threads should be notified
+	 *
+	 * @return true if any virtual threads are notified, otherwise false
+	 */
+	static VMINLINE bool
+	notifyVirtualThread(J9VMThread *vmThread, J9ObjectMonitor *objectMonitor, bool notifyAll)
+	{
+		bool notified = false;
+		J9JavaVM *vm = vmThread->javaVM;
+
+		omrthread_monitor_enter(vm->blockedVirtualThreadsMutex);
+		J9VMContinuation *head = objectMonitor->waitingContinuations;
+		J9VMContinuation *prev = head;
+
+		while ((NULL != head) && (NULL != head->vthread)) {
+			if (J9VMJAVALANGTHREAD_DEADINTERRUPT(vmThread, head->vthread)) {
+				/* Remove virtual threads that have been interrupted. */
+				if (prev == head) {
+					/* This is the first entry. */
+					objectMonitor->waitingContinuations = head->nextWaitingContinuation;
+					head->nextWaitingContinuation = NULL;
+					head = objectMonitor->waitingContinuations;
+					prev = objectMonitor->waitingContinuations;
+				} else {
+					prev->nextWaitingContinuation = head->nextWaitingContinuation;
+					head->nextWaitingContinuation = NULL;
+					head = prev->nextWaitingContinuation;
+				}
+			} else {
+				/* Set the notified and onWaitingList flag for vthread that have not been interrupted. */
+				J9VMJAVALANGVIRTUALTHREAD_SET_NOTIFIED(vmThread, head->vthread, JNI_TRUE);
+				J9VMJAVALANGVIRTUALTHREAD_SET_ONWAITINGLIST(vmThread, head->vthread, JNI_TRUE);
+				notified = true;
+
+				if (!notifyAll) {
+					/* For Object.notify, exit the loop with prev and head pointer set to notified Continuation. */
+					break;
+				}
+
+				prev = head;
+				head = head->nextWaitingContinuation;
+			}
+		}
+
+		if (notified) {
+			/* Move notified virtual threads to the blockedContinuations list for unblocking. */
+			if (notifyAll) {
+				prev->nextWaitingContinuation = vm->blockedContinuations;
```

Comment:
`prev` doesn't point at the first entry in the list (unless it's the only entry).